### PR TITLE
[SPARK-5512][Mllib] Run the PIC algorithm with initial vector suggected by the PIC paper

### PR DESCRIPTION
**Pull request by viirya**
As suggested by the paper of Power Iteration Clustering, it is useful to set the initial vector v0 as the degree vector d. This pr tries to add a running method for that.

### CODE REVIEW

**Comment by viirya** [comment]
Looks like the test failed on an unrelated unit test. Please test it again.

**Comment by mengxr** [comment]
@viirya If that works better than a randomized vector in general, we can replace the current initialization. We set it to a random vector to guarantee that if it far from the first eigenvector. If we want to keep both, instead of adding new method, we can make the switch an option:

~~~
val pic = new PowerIterationClustering()
  .setInitialization("random") // or "degree"
  .run(...)
~~~

**Comment by viirya** [comment]
@mengxr I think it is better to keep both and leave it as an option users can switch.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala [hunk 1]
```diff
@@ -43,10 +43,12 @@ class PowerIterationClusteringModel(
  *
  * @param k Number of clusters.
  * @param maxIterations Maximum number of iterations of the PIC algorithm.
+ * @param initMethod Initialization method.
  */
 class PowerIterationClustering private[clustering] (
     private var k: Int,
-    private var maxIterations: Int) extends Serializable {
+    private var maxIterations: Int,
+    private var initMethod: String = "random") extends Serializable {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala]: We don't put the default values here but in the auxiliary constructor (for Java compatibility when we make it public) Remember to document the default value in the auxiliary constructor.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala [hunk 2]
```diff
@@ -70,6 +72,17 @@ class PowerIterationClustering private[clustering] (
   }
 
   /**
+   * Set the initialization method
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala]: Mention what choices are in the doc.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala [hunk 3]
```diff
@@ -70,6 +72,17 @@ class PowerIterationClustering private[clustering] (
...
+   */
+  def setInitialization(method: String): this.type = {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala]: Shall we change the name to `setInitializationMode` to match `KMeans`?

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala [hunk 4]
```diff
@@ -70,6 +72,17 @@ class PowerIterationClustering private[clustering] (
...
+    this.initMethod = method match {
+      case "random" | "degree" => method
+      case _ => throw new IllegalArgumentException("Incorrect initialization method")
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala]: Include `method` in the error message.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala [hunk 5]
```diff
@@ -149,6 +165,18 @@ private[clustering] object PowerIterationClustering extends Logging {
   }
 
   /**
+   * Generates the degree vector as the vertex properties (v0) to start power iteration.
+   * 
+   * @param g a graph representing the normalized affinity matrix (W)
+   * @return a graph with edges representing W and vertices representing the degree vector
+   */
+  def initDegreeVector(g: Graph[Double, Double]): Graph[Double, Double] = {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala]: Should be private.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala [hunk 6]
```diff
@@ -149,6 +165,18 @@ private[clustering] object PowerIterationClustering extends Logging {
...
+    val sum = g.vertices.values.sum()
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala]: Is it accurate to call the method "degree'? We use the diagonals of `D` but they are not node degrees.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala (hunk 5 above)
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala]: Does it need to be? `randomInit` is not private too. And they are included in a private object `PowerIterationClustering`.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala (hunk 6 above)
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala]: Agree. I have put that in the doc.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala (hunk 5 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala]: No, my bad ...

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks!

**Pull request closed without merging**